crates/hotshot/src/traits/networking/combined_network.rs

```diff
@@ -0,0 +1,487 @@
+        let generator = <CombinedNetworks<
+            TYPES,
+            I,
+            MEMBERSHIP,
+        > as TestableNetworkingImplementation<_, _>>::generator(
+            expected_node_count,
+            num_bootstrap,
+            network_id,
+            da_committee_size,
+            is_da
+        );
+        Box::new(move |node_id| Self {
+            networks: generator(node_id).into(),
+            message_cache: Arc::new(RwLock::new(Cache::new(COMBINED_NETWORK_CACHE_SIZE))),
+            primary_down: Arc::new(AtomicU64::new(0)),
+        })
+    }
+
+    /// Get the number of messages in-flight.
+    ///
+    /// Some implementations will not be able to tell how many messages there are in-flight. These implementations should return `None`.
+    fn in_flight_message_count(&self) -> Option<usize> {
+        None
+    }
+}
+
+#[async_trait]
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>, MEMBERSHIP: Membership<TYPES>>
+    CommunicationChannel<TYPES, Message<TYPES, I>, MEMBERSHIP>
+    for CombinedCommChannel<TYPES, I, MEMBERSHIP>
+{
+    type NETWORK = CombinedNetworks<TYPES, I, MEMBERSHIP>;
+
+    async fn wait_for_ready(&self) {
+        join!(
+            self.primary().wait_for_ready(),
+            self.secondary().wait_for_ready()
+        );
+    }
+
+    async fn is_ready(&self) -> bool {
+        self.primary().is_ready().await && self.secondary().is_ready().await
+    }
+
+    fn shut_down<'a, 'b>(&'a self) -> BoxSyncFuture<'b, ()>
+    where
+        'a: 'b,
+        Self: 'b,
+    {
+        let closure = async move {
+            join!(self.primary().shut_down(), self.secondary().shut_down());
+        };
+        boxed_sync(closure)
+    }
+
+    async fn broadcast_message(
+        &self,
+        message: Message<TYPES, I>,
+        election: &MEMBERSHIP,
+    ) -> Result<(), NetworkError> {
+        let recipients =
+            <MEMBERSHIP as Membership<TYPES>>::get_committee(election, message.get_view_number());
+
+        // broadcast optimistically on both networks, but if the primary network is down, skip it
+        if self.primary_down.load(Ordering::Relaxed) < COMBINED_NETWORK_MIN_PRIMARY_FAILURES
+            || self.primary_down.load(Ordering::Relaxed) % COMBINED_NETWORK_PRIMARY_CHECK_INTERVAL
+                == 0
+        {
+            // broadcast on the primary network as it is not down, or we are checking if it is back up
+            match self
+                .primary()
+                .broadcast_message(message.clone(), recipients.clone())
+                .await
+            {
+                Ok(_) => {
+                    self.primary_down.store(0, Ordering::Relaxed);
+                }
+                Err(e) => {
+                    error!("Error on primary network: {}", e);
+                    self.primary_down.fetch_add(1, Ordering::Relaxed);
+                }
+            };
+        }
+
+        self.secondary()
+            .broadcast_message(message, recipients)
```

Comment:
Long term we probably want to add a topic here. I put down an issue from our conversation earlier this week [here](https://github.com/EspressoSystems/HotShot/issues/1843)